src/box/box.cc

```diff
@@ -4384,6 +4397,10 @@ box_cfg_xc(void)
 	title("running");
 	say_info("ready to accept requests");
 
+	if (is_bootstrap_leader) {
+		box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
```

Comment:
Why not do that somewhere in memtx engine callback instead so that we have all "indexes_built" events in the same file? What about `memtx_engine_bootstrap`?

Reply:
Please see my response in the conversation below. I've moved the whole trigger run sequence for bootstrap leader into `bootstrap_master`.

I don't think we should place INDEXES_BUILT into `memtx_engine_bootstrap` though. This would in turn require placing SNAPSHOT_RECOVERED there, because it should go before INDEXES_BUILT. And we'll have SNAPSHOT_RECOVERED in two files, memtx.cc and box.cc then.

Reply:
> I don't think we should place INDEXES_BUILT into `memtx_engine_bootstrap` though. This would in turn require placing SNAPSHOT_RECOVERED there, because it should go before INDEXES_BUILT.

Should it though? AFAIU "index_built" can be triggered before or after "wal_recovered". What's wrong if it's triggered before "snapshot_recovered"? In case of bootstrap, technically it seems to be correct.

Reply:
> It seems to me we need some order for the triggers (unless we are talking about wal_recovered and indexes_built, which can come in any order).

There's a definite order: snap < wal < sync. The "indexes_built" event is already special: it's < sync, but may be < or > wal. Since it's special, there's nothing wrong in going further and saying that it may be < or > snap. Technically, we could introduce the mode in which indexes are built right before snapshot is recovered, e.g. we could do that in force_recovery mode.

> Maintaining the order seems cheap enough

It's not cheap - it makes the code more difficult to follow and, what is worse, prevents us from building indexes before snapshot is recovered in future.

> For example, @Totktonada suggested that the stages should be comparable in #7563.

Let's go from use cases: suppose the user knows that indexes_built is always fired after snap_recovered. What good does they get from that knowledge? I imagine that the user might want to know the moment when the database becomes usable (everything is recovered, and indexes are built), but currently there's no event for this except 'sync', which isn't quite what they might want; the user is supposed to wait for wal_recovered **and** indexes_built instead, which isn't quite convenient. Moving indexes_built before snap_recovered changes absolutely nothing from that point. Maybe, we should introduce a new event which would be fired when the database becomes fully functional, that is max(snap_recovered, indexes_built)?

Reply:
> Technically, we could introduce the mode in which indexes are built right before snapshot is recovered, e.g. we could do that in force_recovery mode.

Anyway "indexes_built" implies that the data is accessible, beacause it's already recovered and there are indexes for the data.

Building the indexes before recovering the snap makes no sense -- since there is nothing to read yet. Even with "built" indexes.

> the user is supposed to wait for wal_recovered and indexes_built instead

This is true.

> Maybe, we should introduce a new event which would be fired when the database becomes fully functional, that is max(snap_recovered, indexes_built)?

The event you are speaking of should be max(wal_recovered, indexes_built).

After thinking this through, it seems to me we shouldn't care about the real moment when indexes are built, and instead always fire "indexes_built" after "wal_recovered". What do you think?

When indexes are built after local recovery completes, we aren't lying to the user and this stage has a use case: let the user know that the data is fully accessible (via secondary indices).

When indexes are built on top of an intermediate state (snap is recovered, but WAL isn't), there is no use for the user to know that the indexes are built. So it seems placing "indexes_built" after "wal_recovered" is good in this case too.

Besides, the implementation details will be hidden from the user, a full order will be established, and we will be free to build indexes even before the snapshot recovery, as you say.

Reply:
> Building the indexes before recovering the snap makes no sense -- since there is nothing to read yet. Even with "built" indexes.

Why not? This means that indexes are built, but the data isn't recovered yet. As we recover data, we will insert it into all indexes right away.

> After thinking this through, it seems to me we shouldn't care about the real moment when indexes are built, and instead always fire "indexes_built" after "wal_recovered". What do you think?

Sounds reasonable, **but** in the near future, the memtx mvcc mode will be enabled by default and indexes will be always built right after snapshot recovery so indexes_built will be fired right after wal_recovered. Do we even need this event then?

Moreover, thinking even further, in distant future, we may end up with *continuous* snapshotting, i.e. there will be no such thing as snapshot recovery at all and all the three events (snap_recovered, wal_recovered, indexes_built) will be fired one after another without any delay when recovery is fully completed.

Come to think of it, why do we need to separate snapshot_recovered from wal_recovered? Do you have a use case in mind? Why not just add one trigger - on_recovery_complete - which would fire after recovering snap and wal and building indexes?

Reply:
I'll paste @Mons's words here:
```
indexes_built важен тем, что до того как они построены любая триггерная операция потенциально приводит к duplicate_key.
т.е. я вижу indexes_built не как строгий элемент порядка загрузки, а как момент, после которого можно полноценно пользоваться индексами.
```

and 
```
я бы не завязывался на предопределённом порядке
```

So it seems, "indexes_built" should mean what you are saying -- indexes are usable.
And there is a use case for "indexes_built" -- know if constraints are checked already or not (useful for before_replace triggers).

"recovery_finished" (which is wal_recovered now) is also needed: to catch the moment when local data is fully recovered (to access it via primary key at least).

I can't remember why "snapshot_recovered" is needed, but @ochaton asked for it explicitly.

Anyway, seems we don't care about the trigger invocation order for bootstrap. So I'll try to implement your suggestiong with "indexes_built" in `memtx_engine_bootstrap`.

Reply:
> It seems to me we need some order for the triggers (unless we are talking about wal_recovered and indexes_built, which can come in any order). For example, @Totktonada suggested that the stages should be comparable in #7563.
Maintaining the order seems cheap enough (it's the matter of one extra usage of "INDEXES_BUILT" in `bootstrap_master` compared to your solution), so I'd stick with ordered trigger invocation where it is possible.

When I wrote it I thought that the states appear in particular fixed order. If it doesn't reflect the world, let's ignore and go ahead with something that would reflect it better.

Regarding on the rest of the discussion: we can go from use cases (and improve particular user scenarios) or go from stages we already have in the code (improve tarantool's state observability). The latter may solve some use cases as well, but in fact I wouldn't consider it as the primary goal.

Reply:
Done:
```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index afa20ba4f..b75a82c1d 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -3780,7 +3780,6 @@ bootstrap_master(const struct tt_uuid *replicaset_uuid)
 
        box_run_on_recovery_state(RECOVERY_STATE_SNAPSHOT_RECOVERED);
        box_run_on_recovery_state(RECOVERY_STATE_WAL_RECOVERED);
-       box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
        assert(!recovery_state_synced_is_reached);
        box_run_on_recovery_state(RECOVERY_STATE_SYNCED);
        recovery_state_synced_is_reached = true;
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index b5487d620..9a1becaed 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -636,7 +636,10 @@ memtx_engine_bootstrap(struct engine *engine)
                        break;
        }
        xlog_cursor_close(&cursor, false);
-       return rc < 0 ? -1 : 0;
+       if (rc < 0)
+               return -1;
+       box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
+       return 0;
 }
 
 static int
diff --git a/test/replication-luatest/recovery_triggers_test.lua b/test/replication-luatest/recovery_triggers_test.lua
index 755bf2523..f1e9d3050 100644
--- a/test/replication-luatest/recovery_triggers_test.lua
+++ b/test/replication-luatest/recovery_triggers_test.lua
@@ -57,7 +57,7 @@ local function assert_state_not_reached(server, state)
              'State ' .. state .. ' not reached')
 end
 
-local function wait_usual_state_order(server)
+local function wait_all_states(server)
     wait_state(server, 'snapshot_recovered')
     -- wal_recovered and indexes_built can come in any order.
     wait_state(server, 'wal_recovered')
@@ -68,14 +68,14 @@ end
 g.test_recovery_stages = function(cg)
     -- Test stages on bootstrap.
     cg.server:start()
-    wait_usual_state_order(cg.server)
+    wait_all_states(cg.server)
     cg.server:stop()
     -- Test same stages on recovery.
     cg.server:start()
-    wait_usual_state_order(cg.server)
+    wait_all_states(cg.server)
     -- Test the stages on replica join.
     cg.replica:start()
-    wait_usual_state_order(cg.replica)
+    wait_all_states(cg.replica)
     cg.replica:stop()
     -- Test sync state.
     cg.replica.box_cfg = {
```